radiomesh/gridding.py

```diff
@@ -168,32 +184,110 @@ def impl(
           u_pixel_start = int(np.round(u_grid)) - HALF_SUPPORT_INT
           v_pixel_start = int(np.round(v_grid)) - HALF_SUPPORT_INT
 
-          # Tuple of indices associated with each kernel value in X and Y
+          # Tuple of indices associated with each kernel value in X, Y and Z
           # Of length kernel.support
-          x_indices = es_kernel_positions(KERNEL, NX, u_pixel_start)
-          y_indices = es_kernel_positions(KERNEL, NY, v_pixel_start)
+          x_indices = es_kernel_positions(KERNEL, NX, u_pixel_start, True)
+          y_indices = es_kernel_positions(KERNEL, NY, v_pixel_start, True)
 
           # Tuples of kernel values of length kernel.support
           x_kernel = eval_es_kernel(KERNEL, u_grid, u_pixel_start)
           y_kernel = eval_es_kernel(KERNEL, v_grid, v_pixel_start)
 
-          for xfi, xk in zip(
+          for xfi, xkw in zip(
             numba.literal_unroll(x_indices), numba.literal_unroll(x_kernel)
           ):
             xi = int(xfi)
-            for yfi, yk in zip(
+            for yfi, ykw in zip(
               numba.literal_unroll(y_indices), numba.literal_unroll(y_kernel)
             ):
-              pol_weight = xk * yk
               yi = int(yfi)
-              weighted_stokes = apply_weights(vis, pol_weight)
-              accumulate_data(weighted_stokes, vis_grid_view, (xi, yi), 0)
-              # weighted_weights = apply_weights(wgt, pol_weight)
-              # accumulate_data(weighted_weights, weight_grid_view, (xi, yi), 0)
+              weighted_stokes = apply_weights(vis, xkw * ykw)
+              accumulate_data(weighted_stokes, vis_grid, (xi, yi), 0)
+
+    return vis_grid
+
+  def apply_w_impl(
+    uvw,
+    visibilities,
+    weights,
+    flags,
+    frequencies,
+    wgrid_literal_params,
+    jones_params,
+  ):
+    check_args(uvw, visibilities, weights, flags, frequencies, NPOL)
+    ntime, nbl, nchan, _ = visibilities.shape
+
+    wavelengths = frequencies / LIGHTSPEED
+
+    vis_grid = np.zeros((NSTOKES, NW, NX, NY), visibilities.dtype)
+
+    for t in numba.prange(ntime):
+      for bl in range(nbl):
+        u, v, w = load_data(uvw, (t, bl), NUVW, -1)
+        for ch in range(nchan):
+          idx = (t, bl, ch)  # Indexing tuple for use in intrinsics
+          # Return early if any visibility is flagged
+          vis_flag = load_data(flags, idx, NPOL, -1)
+          if reduce(any_flagged, vis_flag, False):
+            continue
+
+          vis = load_data(visibilities, idx, NPOL, -1)
```

Comment:
Why do the jones_params seem constant here? Jones will be `(ntime, nchan, nant, ndir, 2)` or `(ntime, nchan, nant, ndir, 2, 2)` so maybe they should also be fetched with a `load_data`